kmath-core/src/commonMain/kotlin/space/kscience/kmath/operations/LogicAlgebra.kt

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright 2018-2021 KMath contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package space.kscience.kmath.operations
+
+import space.kscience.kmath.expressions.Symbol
+import space.kscience.kmath.expressions.symbol
+import space.kscience.kmath.misc.UnstableKMathAPI
+
+/**
+ * An algebra for generic boolean logic
+ */
+@UnstableKMathAPI
+public interface LogicAlgebra<T : Any> : Algebra<T> {
+
+    /**
+     * Represent constant [Boolean] as [T]
+     */
+    public fun const(boolean: Boolean): T
+
+    override fun bindSymbolOrNull(value: String): T? = value.lowercase().toBooleanStrictOrNull()?.let(::const)
+
+    override fun unaryOperation(operation: String, arg: T): T = when (operation) {
+        Boolean::not.name -> arg.not()
+        else -> super.unaryOperation(operation, arg)
+    }
+
+    override fun unaryOperationFunction(operation: String): (arg: T) -> T = { unaryOperation(operation, it) }
+
+    override fun binaryOperation(operation: String, left: T, right: T): T = when (operation) {
+        Boolean::and.name -> left.and(right)
+        Boolean::or.name -> left.or(right)
+        else -> super.binaryOperation(operation, left, right)
+    }
+
+    override fun binaryOperationFunction(operation: String): (left: T, right: T) -> T = { l, r ->
+        binaryOperation(operation, l, r)
+    }
+
+    /**
+     * Logic 'not'
+     */
+    public operator fun T.not(): T
+
+    /**
+     * Logic 'and'
+     */
+    public infix fun T.and(other: T): T
+
+    /**
+     * Logic 'or'
+     */
+    public infix fun T.or(other: T): T
+
```

Comment:
I think we may add some more logic functions (xor, implies...)

Reply:
They will be added later as extensions

Reply:
There are no multiple receivers on extension functions in Kotlin.

Reply:
Added xor.

Reply:
OK